File: c#/Days/Day2.cs
namespace aoc24.Days;

public static class Day2
{
    public class Input
    {
        public List<List<int>> Reports = [];
    }

    public static Input ParseInput(string s)
    {
        var ret = new Input();
        using var sr = new StringReader(s);
        while (sr.ReadLine() is { } line)
        {
            ret.Reports.Add(line.Split(' ').Select(int.Parse).ToList());
        }
        return ret;
    }

    public static int Part1(Input input)
    {
        return input.Reports.Count(IsValidReport);
    }

    public static int Part2(Input input)
    {
        return input.Reports
            .Count(r => IsValidReport(r) ||
                        Enumerable.Range(0, r.Count)
                            .Any(i =>
                            {
                                var r1 = new List<int>(r);
                                r1.RemoveAt(i);
                                return IsValidReport(r1);
                            }));
    }

    private static bool IsValidReport(List<int> r)
    {
        bool inc = true, dec = true;

        foreach (var (n1, n2) in r.Zip(r.Skip(1)))
        {
            var diff = n2 - n1;
            if (diff <= 0) inc = false;
            if (diff >= 0) dec = false;
            if ((!inc && !dec) || Math.Abs(diff) > 3) return false;
        }

        return true;
    }
}

File: rust/src/days/day2.rs
use std::{convert::Infallible, str::FromStr};

#[derive(Debug)]
struct Input {
    reports: Vec<Vec<i32>>,
}

impl FromStr for Input {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let reports = s
            .lines()
            .map(|line| {
                line.split_ascii_whitespace()
                    .map(|n| n.parse::<i32>().unwrap())
                    .collect()
            })
            .collect();
        Ok(Input { reports })
    }
}

pub fn part1(input: &str) -> usize {
    let input = Input::from_str(input).unwrap();
    input
        .reports
        .into_iter()
        .filter(|report| is_report_valid(report))
        .count()
}

pub fn part2(input: &str) -> usize {
    let input = Input::from_str(input).unwrap();
    input
        .reports
        .into_iter()
        .filter(|report| {
            is_report_valid(report)
                || (0..report.len()).any(|i| {
                    let mut new_report = report.clone();
                    new_report.remove(i);
                    is_report_valid(&new_report)
                })
        })
        .count()
}

fn is_report_valid(report: &Vec<i32>) -> bool {
    let mut inc = true;
    let mut dec = true;
    for pair in report.windows(2) {
        let diff = pair[1] - pair[0];
        if diff > 0 {
            dec = false;
        } else if diff < 0 {
            inc = false;
        }
        if diff.abs() < 1 || diff.abs() > 3 || !dec && !inc {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "\
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9";

    #[test]
    fn test_part1() {
        assert_eq!(2, part1(INPUT));
    }

    #[test]
    fn test_part2() {
        assert_eq!(4, part2(INPUT));
    }
}
